src/analyses/ai_history.h

```diff
@@ -0,0 +1,244 @@
+  ///     POSTCONDITION(IMPLIES(result.first == step_statust::MERGED,
+  ///                           result.second is an element of others));
+  ///  3. Block this flow of execution
+  ///     This indicates the transition is impossible (returning to a location
+  ///     other than the call location) or undesireable (omitting some traces)
+  ///     POSTCONDITION(IMPLIES(result.first == BLOCKED,
+  ///                           result.second == nullptr()));
+  virtual step_returnt step(locationt to, const trace_sett &others) const = 0;
+
+  /// The order for history_sett
+  virtual bool operator<(const ai_history_baset &op) const = 0;
+
+  /// History objects should be comparable
+  virtual bool operator==(const ai_history_baset &op) const = 0;
+
+  /// The current location in the history, used to compute the transformer
+  /// POSTCONDITION(return value is dereferenceable)
+  virtual const locationt &current_location(void) const = 0;
+
+  /// Domains with a substantial height may need to widen when merging
+  /// this method allows the history to provide a hint on when to do this
+  /// The suggested use of this is for domains merge operation to check this
+  /// and then widen in a domain specific way.  However it could be used in
+  /// other ways (such as the transformer).  The key idea is that the history
+  /// tracks / should know when to widen and when to continue.
+  virtual bool should_widen(const ai_history_baset &other) const
+  {
+    return false;
+  }
+
+  virtual void output(std::ostream &out) const
+  {
+  }
+  virtual jsont output_json(void) const;
+  virtual xmlt output_xml(void) const;
+};
+
+/// The common case of history is to only care about where you are now,
+/// not how you got there!
+/// Invariants are not checkable due to C++...
+class ahistoricalt : public ai_history_baset
+{
+protected:
+  // DATA_INVARIANT(current.is_dereferenceable(), "Must not be _::end()")
+  locationt current;
+
+public:
+  explicit ahistoricalt(locationt l) : ai_history_baset(l), current(l)
+  {
+  }
+
+  ahistoricalt(const ahistoricalt &old)
+    : ai_history_baset(old), current(old.current)
+  {
+  }
+
+  step_returnt step(locationt to, const trace_sett &others) const override
+  {
+    trace_ptrt next(new ahistoricalt(to));
+
+    if(others.empty())
+    {
+      return std::make_pair(step_statust::NEW, next);
+    }
+    else
+    {
+      // Aggressively merge histories because they are indistinguishable
+      INVARIANT(others.size() == 1, "Only needs one history per location");
+
+      const auto it = others.begin();
+      INVARIANT(
+        *(*(it)) == *next,
+        "The next location in history map must contain next history");
+
+      return std::make_pair(step_statust::MERGED, *it);
+    }
+  }
+
+  bool operator<(const ai_history_baset &op) const override
+  {
+    PRECONDITION(dynamic_cast<const ahistoricalt *>(&op) != nullptr);
+
+    return this->current_location()->location_number <
```

Comment:
It can and that is intentional.  As I understand it, location numbers are intended to be globally unique within a goto_model.  So comparing location numbers is the safe version of comparing `locationt`s.